enhancements/dns/externaldns-operator-aws-assume-role.md

```diff
@@ -0,0 +1,273 @@
+
+```go
+type ExternalDNSAWSProviderOptions struct {
+    [...]
+    // assumeRole is a reference to the IAM role that
+    // ExternalDNS will be assuming in order to perform
+    // any DNS updates.
+    //
+    // +kubebuilder:validation:Optional
+    // +optional
+    AssumeRole *ExternalDNSAWSAssumeRoleOptions `json:"assumeRole,omitempty"`
+}
+
+type ExternalDNSAWSAssumeRoleOptions struct {
+    // arn is an AWS role ARN that the ExternalDNS
+    // operator will assume when making DNS updates.
+    //
+    // +kubebuilder:validation:Required
+    // +required
+    ARN string `json:"arn,omitempty"`
+}
+
+```
+
+### Implementation Details/Notes/Constraints
+
+#### The Required Credentials Issue
+
+Extending the v1beta1 `ExternalDNS` API object revealed a pre-existing issue with the `Credentials` API field in
+`ExternalDNSAWSProviderOptions`. The problem is that `Credentials` is designated as required to satisfy vanilla
+Kubernetes use cases. On OpenShift, the ExternalDNS Operator generates a `CredentialRequest` automatically, eliminating
+the need to manually specify `Credentials`. This wasn't yet exposed as an issue because there were no other fields in
+`spec.provider.aws`, and `spec.provider.aws` is optional. The underlying [logic](https://github.com/openshift/external-dns-operator/blob/11e3b72b75d696b7419bce8360183373fc0725e1/api/v1beta1/externaldns_webhook.go#L132)
+of the webhook is accurate; the issue lies with this required field in the `ExternalDNS` CRD.
+
+However, there exists a simple workaround: OpenShift users can specify an empty `""` `Credentials` field as follows:
+```yaml
+apiVersion: externaldns.olm.openshift.io/v1beta1
+kind: ExternalDNS
+metadata:
+  name: sample-aws
+spec:
+  provider:
+    type: AWS
+    aws:
+      assumeRole:
+        arn: "arn:aws:xxx"
+      credentials:
+        name: ""
+```
+This satisfies the `ExternalDNS` CRD `+required` field while the ExternalDNS Operator logic [ignores](https://github.com/openshift/external-dns-operator/blob/bac092e98fe5f9065b75bd0fa21d5aeff9d00853/pkg/operator/controller/credentials-secret/controller.go#L259-L267)
+the empty string and uses the `CredentialsRequest`. As a result, we plan to introduce this `AssumeRole` field in the
+v1beta1 release accompanied by documentation that specifies this requirement to specify an empty `Credentials` field in
+OpenShift.
+
+In the next version of ExternalDNS Operator's API, we will address the requirement for the `Credentials` field as
+outlined in the alternative [Bumping the API](#bumping-the-api), thus resolving this particular issue.
+
+### Risks and Mitigations
+
+* Introducing this API change in v1beta1 exposes the issue outlined in [The Required Credentials Issue](#the-required-credentials-issue)
+  and has the potential to cause confusion. We plan to mitigate this by documenting the limitation. 
+
+### Drawbacks
+
+No obvious drawbacks as this enhancement is merely exposing an existing ExternalDNS functionality via an API.
```

Comment:
What if this is misconfigured or hacked for an unwante privilege escalation?  That would be a big drawback.  General security implications aren't discussed here and probably should be - at least to say that it isn't a concern for reasons detailed in the discussion.